### DISCUSSION TITLE
Missing message when using OpenDDS with Java.

### DISCUSSION THREAD

**Discussion by yuan-ge** [OpenDDS/OpenDDS#4319] [Q&A] [answered]
Environment:
* OpenDDS 3.25
* ACE+TAO 7.1
* Java 8
When I run the messenger example in java/tests' publisher and subscriber. I use the rtps file. When I start the publisher then subscriber, it looks correct. But when I start them reversely, the publisher's first message will not be get by subscriber. I can't find the reason, but when I add a sleep before the publisher send the first message, all will be good.
Is this a bug? How can I resolve it?

**Comment by jrw972** [comment]
Are you using reliability?

**Comment by yuan-ge** [comment, marked as answer]
> Are you using reliability?

You give me a hint. I regarded the Qos initialization code as setting code, after setting the reliability, it looks good.
Another question is when will the best effort policy discard the message? The history is set KEEP_ALL, I think the data reader should keep all messages until app read them, even the reliability is setting to BSET_EFFORT.

**Reply by yuan-ge** [reply]
The Java initialization code is so long, that I reagard it as setting code. The setting is done after the `get_default_datawriter_qos()` method is called.

**Comment by jrw972** [comment]
Good question.  It has to do with the fact the discovery is asynchronous and asymmetric.  Logically, one side will complete discovery before the other.  If the publisher completes discovery first, it may send a sample before the subscriber is ready to receive it.

This doesn't happen when using reliability because there is implicit synchronization _based on how it is implemented in OpenDDS_.  I don't think the specs require this behavior, but it is generally useful.  When a reliable writer discovers a reliable reader, it records the current sample sequence number.  Reliability means that every sample after that sample number must be delivered to the subscriber.  Thus, if the reader is not ready when the sample is sent, then the retry mechanism for reliability will cause it to be resent.

**Reply by yuan-ge** [reply]
Thank you for your patient answer.
